Raygun4UWP/RaygunClient.cs

```diff
@@ -0,0 +1,455 @@
+
+            if (InternetAvailable())
+            {
+              await SendCrashReport(payload, true);
+            }
+            else
+            {
+              await SaveCrashReport(payload);
+            }
+          }
+          catch (Exception ex)
+          {
+            Debug.WriteLine($"Error Logging Exception to Raygun: {ex.Message}");
+          }
+        }
+      }
+    }
+
+    private bool ValidateApiKey()
+    {
+      if (string.IsNullOrEmpty(Settings.ApiKey))
+      {
+        System.Diagnostics.Debug.WriteLine("ApiKey has not been provided, exception will not be logged");
+        return false;
+      }
+
+      return true;
+    }
+
+    // Returns true if the crash report can be sent, false if the sending is canceled.
+    private bool OnSendingCrashReport(Exception originalException, RaygunCrashReport raygunCrashReport)
+    {
+      bool result = true;
+
+      if (!_handlingRecursiveErrorSending)
+      {
+        EventHandler<RaygunSendingCrashReportEventArgs> handler = SendingCrashReport;
+        if (handler != null)
+        {
+          RaygunSendingCrashReportEventArgs args = new RaygunSendingCrashReportEventArgs(originalException, raygunCrashReport);
+          try
+          {
+            handler(this, args);
+          }
+          catch (Exception e)
+          {
+            // Catch and send exceptions that occur in the SendingCrashReport event handler.
+            // Set the _handlingRecursiveErrorSending flag to prevent infinite errors.
+            _handlingRecursiveErrorSending = true;
+            Send(e);
+            _handlingRecursiveErrorSending = false;
+          }
+
+          result = !args.Cancel;
+        }
+      }
+
+      return result;
+    }
+
+    private async void BeginSendStoredCrashReports()
+    {
+      await SendStoredCrashReports();
+    }
+
+    private async Task SendStoredCrashReports()
+    {
+      if (InternetAvailable())
+      {
+        try
+        {
+          var tempFolder = ApplicationData.Current.TemporaryFolder;
+
+          var raygunFolder = await tempFolder.CreateFolderAsync(OFFLINE_DATA_FOLDER, CreationCollisionOption.OpenIfExists).AsTask().ConfigureAwait(false);
+
+          var files = await raygunFolder.GetFilesAsync().AsTask().ConfigureAwait(false);
+
+          foreach (var file in files)
+          {
+            string text = await FileIO.ReadTextAsync(file).AsTask().ConfigureAwait(false);
+            await SendCrashReport(text, false).ConfigureAwait(false);
+
+            await file.DeleteAsync().AsTask().ConfigureAwait(false);
+          }
+
+          await raygunFolder.DeleteAsync().AsTask().ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+          Debug.WriteLine($"Error sending stored crash reports to Raygun: {ex.Message}");
+        }
+      }
+    }
+
+    private async Task SendCrashReport(string payload, bool saveOnFail)
+    {
+      var httpClient = new HttpClient();

```

Comment:
We shouldn't be creating a new HttpClient every time we send a report. Instead there should be a single static HttpClient instance for the RaygunClient instance. See https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=dotnet-uwp-10.0
> HttpClient is intended to be instantiated once and re-used throughout the life of an application. Instantiating an HttpClient class for every request will exhaust the number of sockets available under heavy loads.

Reply:
Have done this now by putting common request logic into a static HttpService class that will hold a singleton HttpClient.